numpy/core/src/umath/string_ufuncs.cpp

```diff
@@ -0,0 +1,457 @@
+            if (cmp != 0) {
+                return cmp;
+            }
+            str1++;
+            str2++;
+        }
+    }
+    if (len1 > len2) {
+        for (int i = n; i < len1; i++) {
+            int cmp = character_cmp(*str1, (character)0);
+            if (cmp != 0) {
+                return cmp;
+            }
+            str1++;
+        }
+    }
+    else if (len2 > len1) {
+        for (int i = n; i < len2; i++) {
+            int cmp = character_cmp((character)0, *str2);
+            if (cmp != 0) {
+                return cmp;
+            }
+            str2++;
+        }
+    }
+    return 0;
+}
+
+
+template <bool rstrip, int comp, typename character>
+static int
+string_comparison_loop(PyArrayMethod_Context *context,
+        char *const data[], npy_intp const dimensions[],
+        npy_intp const strides[], NpyAuxData *NPY_UNUSED(auxdata))
+{
+    /*
+     * Note, fetching `elsize` from the descriptor is OK even without the GIL,
+     * however it may be that this should be moved into `auxdata` eventually,
+     * which may also be slightly faster/cleaner (but more involved).
+     */
+    int len1 = context->descriptors[0]->elsize / sizeof(character);
+    int len2 = context->descriptors[1]->elsize / sizeof(character);
+
+    char *in1 = data[0];
+    char *in2 = data[1];
+    char *out = data[2];
+
+    npy_intp N = dimensions[0];
+
+    while (N--) {
+        int cmp = string_cmp<rstrip>(
+                len1, (character *)in1, len2, (character *)in2);
+        npy_bool res;
+        switch (comp) {
+            case Py_EQ:
+                res = cmp == 0;
+                break;
+            case Py_NE:
+                res = cmp != 0;
+                break;
+            case Py_LT:
+                res = cmp < 0;
+                break;
+            case Py_LE:
+                res = cmp <= 0;
+                break;
+            case Py_GT:
+                res = cmp > 0;
+                break;
+            case Py_GE:
+                res = cmp >= 0;
+                break;
+            default:
+                assert(false);
+        }
+        *(npy_bool *)out = res;
+
+        in1 += strides[0];
+        in2 += strides[1];
+        out += strides[2];
+    }
+    return 0;
+}
+
+
+/*
+ * Machinery to add the string loops to the existing ufuncs.
+ */
+
+/*
+ * This function replaces the strided loop with the passed in one,
+ * and registers it with the given ufunc.
+ */
+static int
+add_loop(PyObject *umath, const char *ufunc_name,
+         PyArrayMethod_Spec *spec, PyArrayMethod_StridedLoop *loop)
+{
+    PyObject *name = PyUnicode_FromString(ufunc_name);
+    if (name == nullptr) {
+        return -1;
+    }
+    PyObject *ufunc = PyObject_GetItem(umath, name);
+    Py_DECREF(name);
+    if (ufunc == nullptr) {
+        return -1;
+    }
+    spec->slots[0].pfunc = (void *)loop;
+
+    int res = PyUFunc_AddLoopFromSpec(ufunc, spec);
+    Py_DECREF(ufunc);
+    return res;
+}
+
+
+extern "C" {
```

Comment:
Like below: I think it makes sense to create a 'strings_ufunc.h' that contains these definitions.

Reply:
Maybe @mattip, you have a good idea?  I had that first, and it should work for the `init_string_ufuncs`, but the umath/multiarray boundary is always creating some headaches: it seems hard to include that header from `multiarray` somehow?

I admit, it is very likely I am just doing something stupid.

Also @serge-sans-paille, I thought another problem was that the symbol export didn't actually work right when I had it as its own header here.  But I guess that may have been just a bad try?

Reply:
Maybe good to separate what should more obviously work, which is to put the ones double-defined here in 'strings_ufunc.h' and import that new .h above. And only then worry about the other places. 

Regarding multiarray/umath, I thought we had tried to harmonize those, but not sure...

Reply:
Yes, they are now one build-unit, which is why this works at all without much hassle.  But including umath stuff in multiarray is still tricky (I am not exactly sure why)

Reply:
OK, moved things back, it is working fine, but did not try to cross the `multiarray` boundary yet.  I guess one thing could be that I did not actually include other stuff in the header right now (not sure if I should).